File: julekalender/2024/knowit/luke03/main.hs
import qualified Data.Map as Map

-- Name, Amount, Number eaten, Seconds since eaten
type Food = [(String, Int, Int, Int)]

type Refill = Map.Map String [Int]

foodTypes = ["ris", "erter", "gulrøtter", "reinsdyrkjøtt", "julekringle"]

food :: Food
food = map (,100,0,0) foodTypes

foodRefill :: Refill
foodRefill =
  Map.fromList
    [ ("ris", [0, 0, 1, 0, 0, 2]),
      ("erter", [0, 3, 0, 0]),
      ("gulrøtter", [0, 1, 0, 0, 0, 8]),
      ("reinsdyrkjøtt", [100, 80, 40, 20, 10]),
      ("julekringle", [0])
    ]

getFood :: Food -> String -> Int
getFood [] _ = 0
getFood ((name, amount, _, _) : food) target
  | name == target = amount
  | otherwise = getFood food target

refillFood :: Food -> Refill -> Int -> Food
refillFood [] _ _ = []
refillFood food@((name, amount, n, s) : xs) refill seconds
  | name == "gulrøtter" = if seconds >= 30 then (name, amount + refillAmount, n + 1, s + 1) : refillFood xs refill seconds else (name, amount, n, s + 1) : refillFood xs refill seconds
  | name == "reinsdyrkjøtt" = if amount <= 0 && s > 50 && n < 5 then (name, amount + refillAmount, n + 1, 0) : refillFood xs refill seconds else (name, amount, n, s + 1) : refillFood xs refill seconds
  | name == "julekringle" = (name, amount, n, s + 1) : refillFood xs refill seconds
  | otherwise = (name, amount + refillAmount, n + 1, s + 1) : refillFood xs refill seconds
  where
    refillAmount = case Map.lookup name refill of
      Just x -> x !! (n `mod` length x)
      Nothing -> 0

eatTwo' :: [Int] -> Int -> Food -> Food
eatTwo' _ _ [] = []
eatTwo' [] _ x = x
eatTwo' g@(gram : grams) eaten x@((name, amount, n, s) : xs)
  | eaten >= 2 = (name, amount, n, s + 1) : eatTwo' g eaten xs
  | name == "reinsdyrkjøtt" = if amount >= 2 && eaten == 0 then (name, max 0 (amount - 2), n, s) : eatTwo' grams (eaten + 2) xs else next
  | name == "julekringle" = if amount >= 1 && eaten == 0 then (name, max 0 (amount - 1), n, s) : eatTwo' grams (eaten + 2) xs else next
  | amount >= gram = (name, max 0 (amount - gram), n, s) : eatTwo' grams (eaten + 1) xs
  | otherwise = next
  where
    next = (name, amount, n, s) : eatTwo' g eaten xs

eatTwo :: Food -> Food
eatTwo = eatTwo' [5, 3] 0

eatAndRefill :: Food -> Refill -> Int -> Food
eatAndRefill food = refillFood (eatTwo food)

eatAndRefill' food = eatAndRefill food foodRefill 0

refillAndEat food refill seconds = eatTwo (refillFood food refill seconds)

solve :: Food -> Refill -> Int -> Int
solve food refill seconds
  | newFood `getFood` "julekringle" <= 99 = seconds + 1
  | otherwise = solve newFood refill (seconds + 1)
  where
    newFood = refillAndEat food refill seconds

main :: IO ()
main = do
  -- let test = iterate eatTwo food !! 40
  -- print food
  -- print test
  -- print $ refillFood food foodRefill 0
  print food
  -- print $ eatAndRefill food foodRefill 0
  -- print $ eatAndRefill (eatAndRefill food foodRefill 0) foodRefill 1
  -- print $ eatAndRefill (eatAndRefill (eatAndRefill food foodRefill 0) foodRefill 1) foodRefill 2
  print $ eatAndRefill' food
  print $ iterate eatAndRefill' food !! 22
  print $ iterate eatAndRefill' food !! 23
  print $ iterate eatAndRefill' food !! 24
  -- print $ take 70 $ iterate eatAndRefill' food
  let answer = solve food foodRefill 0
  print answer

File: julekalender/2024/knowit/luke03/main.py
class Tallerken:
    def __init__(self):
        self.food = {
            "ris": 100,
            "erter": 100,
            "gulrøtter": 100,
            "reinsdyrkjøtt": 100,
            "julekringle": 100,
        }

        self.refill = {
            "ris": [0, 0, 1, 0, 0, 2],
            "erter": [0, 3, 0, 0],
            "gulrøtter": [0, 1, 0, 0, 0, 8],
            "reinsdyrkjøtt": [100, 80, 40, 20, 10],
            "julekringle": [0],
        }

        self.refill_status = {
            "ris": 0,
            "erter": 0,
            "gulrøtter": 0,
            "reinsdyrkjøtt": 0,
            "julekringle": 0,
        }

        self.seconds = 0
        self.seconds_since_reinsdyrkjøtt = 0

    def eat_food(self):
        eaten = 0
        for food, amount in self.food.items():
            if food == "reinsdyrkjøtt" and amount >= 2:
                if eaten == 0:
                    self.food[food] -= 2
                    break
            elif food == "julekringle" and amount >= 1:
                if eaten == 0:
                    self.food[food] -= 1
                    break
            elif food != "reinsdyrkjøtt" and food != "julekringle":
                grams_to_eat = 5 if eaten == 0 else 3
                if amount >= grams_to_eat:
                    self.food[food] -= grams_to_eat
                    eaten += 1
                if eaten >= 2:
                    break

    def refill_food(self):
        for food, status in self.refill_status.items():
            refills = self.refill[food]
            next_refill = refills[status % len(refills)]

            if food == "gulrøtter":
                if self.seconds >= 30:
                    self.food[food] += next_refill
                    self.refill_status[food] += 1
            elif food == "reinsdyrkjøtt":
                if self.food[food] <= 0:
                    if self.seconds_since_reinsdyrkjøtt >= 50 and status < len(refills):
                        self.food[food] += next_refill
                        self.seconds_since_reinsdyrkjøtt = 0
                        self.refill_status[food] += 1
                    else:
                        self.seconds_since_reinsdyrkjøtt += 1
            elif food == "julekringle":
                pass
            else:
                self.food[food] += next_refill
                self.refill_status[food] += 1

    def tick(self):
        self.eat_food()
        self.refill_food()
        self.seconds += 1


if __name__ == "__main__":
    tallerken = Tallerken()
    while tallerken.food["julekringle"] > 99:
        tallerken.tick()
        print(tallerken.seconds)
        print(tallerken.food)
    print(tallerken.seconds)
